tensorflow/contrib/rnn/python/ops/rnn_cell.py

```diff
@@ -1330,6 +1330,151 @@ def assert_shape_match(inp, out):
     return (res_outputs, new_state)
 
 
+class LayerNormBasicGRUCell(rnn_cell_impl._LayerRNNCell):
+  """GRU unit with layer normalization.
+
+    This class adds layer normalization to a
+    basic GRU unit. Layer normalization implementation is based on:
+
+      https://arxiv.org/abs/1607.06450.
+
+    "Layer Normalization"
+    Jimmy Lei Ba, Jamie Ryan Kiros, Geoffrey E. Hinton
+
+    and is applied before the internal nonlinearities.
+    """
+
+  def __init__(self, num_units,
+               activation=math_ops.tanh,
+               layer_norm=True,
+               norm_gain=1.0,
+               norm_shift=0.0,
+               reuse=None):
+    """Initializes the cell.
+
+        Args:
+          num_units: int, The number of units in the GRU cell.
+          activation: Activation function of the inner states.
+          layer_norm: If `True`, layer normalization will be applied.
+          norm_gain: float, The layer normalization gain initial value. If
+            `layer_norm` has been set to `False`, this argument will be ignored.
+          norm_shift: float, The layer normalization shift initial value. If
+            `layer_norm` has been set to `False`, this argument will be ignored.
+        """
+
+    super(LayerNormBasicGRUCell, self).__init__(_reuse=reuse)
+
+    self._num_units = num_units
+    self._activation = activation
+    self._layer_norm = layer_norm
+    self._g = norm_gain
+    self._b = norm_shift
+    self._reuse = reuse
+
+  @property
+  def state_size(self):
+    return self._num_units
+
+  @property
+  def output_size(self):
+    return self._num_units
+
+  def _linear(self, args, kernel, bias):
+    out = math_ops.matmul(args, kernel)
+    if not self._layer_norm:
+      out = nn_ops.bias_add(out, bias)
+    return out
+
+  def _norm(self, inp, scope):
+    return layers.layer_norm(inp, reuse=True, scope=scope)
+
+  def build(self, inputs_shape):
+    if inputs_shape[1].value is None:
+
+      raise ValueError("Expected inputs.shape[-1] to be known, saw shape: %s"
+                       % inputs_shape)
+
+    input_depth = inputs_shape[1].value
+
+    # Initialize beta and gamma for use by layer_norm.
+    if self._layer_norm:
+      scopes = ["update_gate",
+                "reset_gate",
+                "candidate_linear_x",
+                "candidate_linear_h"]
+      for scope in scopes:
+        self.add_variable(scope+"/gamma",
```

Comment:
nits: place spaces around the "+" here and elsewhere

Reply:
done